File: apps/api/internal/models/activity_log.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

// ActivityLog records every successful authenticated mutation, with a
// tamper-evident hash chain — each row's Hash is SHA-256 of (PrevHash
// || canonical(this_row)). Mutating any row breaks the chain on the
// next VerifyChain pass.
//
// The payload digest is a SHA-256 of the request body so we have
// evidence of what was sent without storing PII verbatim. Read-only —
// no updates, no deletes (use a separate retention job to prune old
// rows; deletion still breaks the chain so it must rebuild from a
// safe checkpoint).
type ActivityLog struct {
	ID            string    `gorm:"primarykey;size:36" json:"id"`
	UserID        string    `gorm:"size:36;index" json:"user_id"`
	Method        string    `gorm:"size:10" json:"method"`
	Path          string    `gorm:"size:500;index" json:"path"`
	Status        int       `json:"status"`
	PayloadDigest string    `gorm:"size:64" json:"payload_digest"` // sha256 hex
	IPAddress     string    `gorm:"size:45" json:"ip_address"`
	UserAgent     string    `gorm:"size:500" json:"user_agent"`
	DurationMS    int64     `json:"duration_ms"`
	PrevHash      string    `gorm:"size:64" json:"prev_hash"` // hex sha256, "" for the genesis row
	Hash          string    `gorm:"size:64;uniqueIndex" json:"hash"` // hex sha256(prev_hash || canonical)
	CreatedAt     time.Time `gorm:"index" json:"created_at"`
}

func (a *ActivityLog) BeforeCreate(tx *gorm.DB) error {
	if a.ID == "" {
		a.ID = uuid.New().String()
	}
	return nil
}

File: packages/shared/schemas/activity-log.ts
import { z } from "zod";

export const CreateActivityLogSchema = z.object({
  userId: z.string(),
  method: z.string(),
  path: z.string(),
  status: z.number().int(),
  payloadDigest: z.string(),
  ipAddress: z.string(),
  userAgent: z.string(),
  durationMs: z.number().int(),
  prevHash: z.string(),
  hash: z.string(),
});

export const UpdateActivityLogSchema = z.object({
  userId: z.string().optional(),
  method: z.string().optional(),
  path: z.string().optional(),
  status: z.number().int().optional(),
  payloadDigest: z.string().optional(),
  ipAddress: z.string().optional(),
  userAgent: z.string().optional(),
  durationMs: z.number().int().optional(),
  prevHash: z.string().optional(),
  hash: z.string().optional(),
});

export type CreateActivityLogInput = z.infer<typeof CreateActivityLogSchema>;
export type UpdateActivityLogInput = z.infer<typeof UpdateActivityLogSchema>;
